src/coreclr/vm/CachedInterfaceDispatchPal.h

```diff
@@ -0,0 +1,191 @@
+
+    uintptr_t GetVTableOffset() const
+    {
+        if (CellType == DispatchCellType::VTableOffset)
+        {
+            uint32_t slot = Token.GetSlotNumber();
+            unsigned offsetOfIndirection = MethodTable::GetVtableOffset() + MethodTable::GetIndexOfVtableIndirection(slot) * TARGET_POINTER_SIZE;
+            unsigned offsetAfterIndirection = MethodTable::GetIndexAfterVtableIndirection(slot) * TARGET_POINTER_SIZE;
+
+            uintptr_t offsetOfIndirectionPortion = (((uintptr_t)offsetOfIndirection) << ((TARGET_POINTER_SIZE * 8) / 2));
+            uintptr_t offsetAfterIndirectionPortion = (((uintptr_t)offsetAfterIndirection) << ((TARGET_POINTER_SIZE * 8) / 4));
+            uintptr_t flagPortion = (uintptr_t)IDC_CachePointerPointsIsVTableOffset;
+
+            uintptr_t result = offsetOfIndirectionPortion | offsetAfterIndirectionPortion | flagPortion;
+            _ASSERTE(slot == VTableOffsetToSlot(result));
+            return result;
+        }
+        return 0;
+    }
+
+    static unsigned VTableOffsetToSlot(uintptr_t vtableOffset)
+    {
+        unsigned offsetOfIndirection = (unsigned)(vtableOffset >> ((TARGET_POINTER_SIZE * 8) / 2));
+        unsigned offsetAfterIndirection = (unsigned)(vtableOffset >> ((TARGET_POINTER_SIZE * 8) / 4)) & 0xFF;
+        unsigned slotGroupPerChunk = (offsetOfIndirection - MethodTable::GetVtableOffset()) / TARGET_POINTER_SIZE;
+        unsigned slot = (slotGroupPerChunk * VTABLE_SLOTS_PER_CHUNK) + (offsetAfterIndirection / TARGET_POINTER_SIZE);
+        return slot;
+    }
+
+    const uint8_t HasCache = 0;
+};
+
+struct InterfaceDispatchCacheHeader
+{
+private:
+    enum Flags
+    {
+        CH_TypeAndSlotIndex = 0x0,
+        CH_MetadataToken = 0x1,
+        CH_Mask = 0x3,
+        CH_Shift = 0x2,
+    };
+
+public:
+    void Initialize(DispatchToken token)
+    {
+        m_token = token;
+    }
+
+    void Initialize(const DispatchCellInfo *pNewCellInfo)
+    {
+        m_token = pNewCellInfo->Token;
+    }
+
+    DispatchCellInfo GetDispatchCellInfo()
+    {
+        DispatchCellInfo cellInfo(m_token, true);
+        return cellInfo;
+    }
+
+private:
+    DispatchToken m_token;
+    TADDR padding; // Ensure that the size of this structure is a multiple of 2 pointers
+};
+
+// One of these is allocated per interface call site. It holds the stub to call, data to pass to that stub
+// (cache information) and the interface contract, i.e. the interface type and slot being called.
+struct InterfaceDispatchCell
+{
+    // The first two fields must remain together and at the beginning of the structure. This is due to the
+    // synchronization requirements of the code that updates these at runtime and the instructions generated
+    // by the binder for interface call sites.
+    TADDR      m_pStub;    // Call this code to execute the interface dispatch
+    volatile TADDR m_pCache;   // Context used by the stub above (one or both of the low two bits are set
```

Comment:
I thought we were supposed to be using `Volatile<TADDR>` instead of the `volatile` keyword. I prefer the keyword personally, but I'd like to ensure we are using it here in an expected manner.

Reply:
Eh...  this code is actually written to C++ volatile semantics, not the semantics of Volatile<T>, but Volatile<T> is a stronger requirement, so I'm happy enough to use it.